File: CrackingTheCodingInterview/Chapter1/one-four.py
# palindrome permutation

def check(string):
    string = string.lower()
    a_value = ord("a")
    z_value = ord("z")
    array = [0]*26
    true_length = 0
    for x in range(0, len(string)):
        value = ord(string[x])
        if(value >= a_value and value <= z_value):
            array[value-a_value] += 1
            true_length += 1
    odd = 0
    for value in array:
        if (value % 2 == 1):
            odd += 1
    if (true_length % 2):
        if(odd == 1):
            return True
        else:
            return False
    else:
        if(odd):
            return False
    return True

print(check("tact coa")) #true
print(check("h e  h")) #true
print(check("hh"))  #true
print(check("e")) #true
print(check("abc")) #false

File: CrackingTheCodingInterview/Chapter1/OneFour.java
public class OneFour {

    public static void main(String[] args){
        System.out.println(check("taco cat"));
        System.out.println(check("h e h"));
        System.out.println(check("hh"));
        System.out.println(check("h"));
        System.out.println(check("abc"));
        
    }

    static Boolean check(String str){
        String lowerCase = str.toLowerCase();
        int[] array = new int[26];
        int a_value = Character.valueOf('a');
        int z_value = Character.valueOf('z');
        int trueLength = 0;
        for (int i = 0; i < lowerCase.length(); i++){
            int value = Character.valueOf(lowerCase.charAt(i));
            if (value >= a_value && value <= z_value){
                array[value-a_value]++;
                trueLength++;
            }
        }
        int odd = 0;
        for (int i : array){
            if(i % 2 == 1){
                odd++;
            }
        }
        if (trueLength % 2 == 0){
            if (odd == 0){
                return true;
            }
            else{
                return false;
            }
        }
        else{
            if (odd == 1){
                return true;
            }
        }

        return false;
    }
    
}